onnx/onnx-ml.proto

```diff
@@ -754,6 +775,61 @@ message OperatorSetIdProto {
   optional int64 version = 2;
 }
 
+// Operator/function status.
+enum OperatorStatus {
+    EXPERIMENTAL = 0;
+    STABLE = 1;
+}
+
+message FunctionProto {
+  // The name of the function, similar usage of op_type in OperatorProto.
+  // Combined with FunctionProto.domain, this forms the unique identity of
+  // the FunctionProto.
+  optional string name = 1;
+
+  // Deprecated since IR Verison 8
+  // optional int64 since_version = 2;
+  reserved 2;
+  reserved "since_version";
+
+  // Deprecated since IR Verison 8
+  // optional OperatorStatus status = 3;
+  reserved 3;
+  reserved "status";
+
+  // The inputs and outputs of the function.
+  repeated string input = 4;
+  repeated string output = 5;
+
+  // The attributes of the function.
+  repeated string attribute = 6;
+
+  // The nodes in the function.
+  repeated NodeProto node = 7;
+  // A human-readable documentation for this function. Markdown is allowed.
+  optional string doc_string = 8;
+
+  // The OperatorSets this function body (graph) relies on.
+  // All FunctionProtos implicitly import the operator set which the function operator is part of.
+  //
+  // All nodes in the function body (graph) will bind against the operator
+  // with the same-domain/same-op_type operator with the HIGHEST version
+  // in the referenced operator sets. This means at most one version can be relied
+  // for one domain.
+  // 
+  // The opserator sets imported by FunctionProto should be compatible with the ones
+  // imported by ModelProto. Example, if same operator set say 'A' is imported by FunctionProto
+  // and ModelProto then versions for the operator set may be different but, 
```

Comment:
I'm struggling to understand why allowing different versions is useful here. I could probably contrive two functions that imported different opset versions for a domain without using ops that are different across the versions, but this seems like it would be fragile.

Would it be better to be more restrictive at first (all versions must be equal for a given domain) and then relax it in a later version?

Reply:
restricting the version to be same as the version imported by model will not work for FunctionOps. Consider this example: Function Foo belongs to domain custom_domain and imports op "Add" from onnx domain in its function body... At the time of authoring this function "Foo", onnx latest released opset version was 13 so Foo imports onnx domain opset 13. Now, with this restriction on opset version the function Foo can only be used in models which import opset 13.... later opset wont work even if schema for Add remains same... Allowing different yet compatible version will allow models of higher opset to use function Foo.

Reply:
Thanks for clarifying! I think there might be a more precise way to state this. Something like:

> The operator sets imported by a FunctionProto must be "compatible" with the ones imported by the ModelProto that invokes them. Two operator sets are compatible if (i) they have the same domain and version, or (ii) they have the same domain and different versions, and all operators from the earlier version are present, with the same schema, in the later version. Intuitively condition (ii) means that a graph using the earlier version of a domain can be transparently "upgraded" to the later version by changing the version number.

I might be mangling the distinction between operator set and domain-version pair here. I'm also not sure if compatibility between model/function imports should be symmetric - as in the text here - or asymmetric. I think the checker implementation is asymmetric, and that might make sense, since IIUC we'd only consider "upgrading" a function (by inlining it into a model), and not a model itself.

Reply:
can you elaborate on symmetric vs asymmetric compatibility

Reply:
Symmetric would be more permissive. e.g. It would be legal to have model-imports-ONNX-10 and function-imports-ONNX-9 **or** model-imports-ONNX-9 and function-imports-ONNX-10. 

Whereas an asymmetric policy would only allow the larger version number in one case... most likely the model. This approach might make sense for an implementation based on inlining, because you could implement this as "upgrading" the function when it is inlined.

IIUC the logic here is asymmetric:

https://github.com/onnx/onnx/pull/3532/files/8a4cdd1d04c002f8db20e2a3ec32a0258515d99c#diff-27b8667be1a3c2c7ce2a2d324700c2f7bf400916c6b0e3508c8b27f85399ac6aR748

I suppose you could "downgrade" the function to an earlier opset when it is inlined, if you wanted to support a symmetric policy. But this doesn't seem necessary for the example you gave in your comment.

Reply:
We want asymmetric compatibility and not symmetric. This is in sync with what we do for operators as well... A model can import a version <= model opset import version but never the other way...

Reply:
That makes sense. In that case, we'll definitely need to change the comment to make that clear: the current "the versions for the operator set may be different" implies that the function version > model version is allowed.

Reply:
I am unclear why we need asymmetric compatibility. Symmetric ("downgrading" the function as Derek says) would work just as well, without imposing any extra work or challenges, right? That way, when I define a function, I don't need to go to the trouble of finding the earliest opset it will work with ... I define it for the current opset, and transparent downgrading makes it as general as possible.

Reply:
When defining a function the author needs to import the domain and version which they are using for defining the function body. For example if the function body contains op 'A' from domain 'X' then the author should import {'X', 'then_latest_ version'} in the function body. This way any model which imports version equal to or greater than the one imported by the function proto should work. If the schema for op 'A' changes then the function op needs to be updated with the new schema for op 'A' and corresponding import.

Reply:
The author has several choices in above setting: import {'X', 'latest version of X'} (assuming this contains op 'A') or import {'X', 'earliest version of X that contains the corresponding op 'A''}. Since the opset-compatibility test can figure all this out, why do we need the extra constraint 'model which imports version equal to or greater than the one imported by the function proto' ?

Reply:
import {'X', 'latest version of X'} (assuming this contains op 'A') => if <latest version> -1 contains op A then <latets version> will also contain op A. I dont see any motivation for the author to use anything other than latest version... By using a version earlier than latest version you restrict the usage of the function. 
Opset compatibility test only validates whether opsets are compatible... the upgrades are automatic meaning there is no schema change because the schema has not changed.

Reply:
after our discussion I updated the checker to reflect symmetric compatibility